### PR TITLE
#136 server ips

### LINKED ISSUE #136: Server: IPs
Entwicklungsplan page 56
- get and update IP data
- get emails for the team (and the QMs)

### PR DESCRIPTION
**Pull request by arisb1470**
Added retrieveIP, updateIP and retrieveTeamMails

Tests (on Thunderclient) -

retrieveIP:
GET https://localhost:3030/trainees/ip/62

update IP:
PATCH https://localhost:3030/trainees/ip/62
Body (an example)
{
      "generationID": 3,
      "projektname": "test", 
      "kuerzel": "ts",
      "kickoff": "2000-12-12",
      "AngebotBeiEV": true,
      "ZPBeiEV": true,
      "ZPGehalten": "2000-12-12",
      "APBeiEV": true,
      "APGehalten": "2000-12-12",
      "DLBeiEV": true
      
}

retrieveTeamMails:
GET https://localhost:3030/trainees/ip/62/mails
IMPORTANT: To test this method, you need to create a new member in the database, as there are currently no existing trainees. That member needs to have the following content: mitgliedstatus = 1, internesprojekt = 62 and jbt_email shouldn't be NULL. 

Closes #136

### CODE REVIEW

File: server/src/trainees/traineesController.ts [hunk 1]
```diff
@@ -1,11 +1,39 @@
...
 import database = require("../database");
-import * as traineeTypes from "./traineesTypes";
+
+import { Request, Response } from "express";
+import * as traineesTypes from "./traineesTypes";
+import { doesPermissionsInclude } from "../utils/authUtils";
 import { checkForSQLKeywords } from "../utils/stringUtils";
 
+
+/**
```
Comment (BorchBe99) [inline server/src/trainees/traineesController.ts:12]: Explain the input values like "@param {TYPE} name description"

File: server/src/trainees/traineesController.ts [hunk 2]
```diff
@@ -34,6 +62,81 @@ export const retrieveTraineeChoice = (req: Request, res: Response): void => {
 };
 
 /**
```
Comment (BorchBe99) [inline server/src/trainees/traineesController.ts:64]: Explain the input values like "@param {TYPE} name description"

File: server/src/trainees/traineesController.ts [hunk 3]
```diff
@@ -34,6 +62,81 @@ export const retrieveTraineeChoice = (req: Request, res: Response): void => {
...
+      })
+      .catch((err) => {
+        res.status(500).send("Query Error: Updating IP failed");
+      });
+  } else {
+    res.status(403).send("Authorization failed: You are not permitted to do this");
+  }
+};
+
+/**
```
Comment (BorchBe99) [inline server/src/trainees/traineesController.ts:116]: Explain the input values like "@param {TYPE} name description"

File: server/src/trainees/traineesController.ts [hunk 4]
```diff
@@ -34,6 +62,81 @@ export const retrieveTraineeChoice = (req: Request, res: Response): void => {
...
+ * Retrieves the mails for the specified internal projects
+ */
+export const retrieveTeamMails = (req: Request, res: Response) => {
+  database
+    .query(
+      `SELECT jbt_email
```
Comment (BorchBe99) [inline server/src/trainees/traineesController.ts:122]: For the query use an inner join like:

```
SELECT jbt_email
FROM mitglied
INNER JOIN internesprojekt
ON mitglied.internesprojekt = internesprojekt.internesprojektID
WHERE internesProjektID = ?;
```

**Review by BorchBe99 — COMMENTED** [review]
Great job!👍 Only minor changes necessary.

File: server/src/trainees/traineesController.ts (hunk 1 above)
Comment (LennartRukower) [reply server/src/trainees/traineesController.ts:12]: We almost never did this in the members controller. Do you think this is necessary? These are basic request and response types of express.

When performing the backend refactoring, we can think about, what part of the code should be documented. What do you think?

File: server/src/trainees/traineesController.ts (hunk 2 above)
Comment (LennartRukower) [reply server/src/trainees/traineesController.ts:64]: We almost never did this in the members controller. Do you think this is necessary? These are basic request and response types of express.

When performing the backend refactoring, we can think about, what part of the code should be documented. What do you think?

File: server/src/trainees/traineesController.ts (hunk 3 above)
Comment (LennartRukower) [reply server/src/trainees/traineesController.ts:116]: We almost never did this in the members controller. Do you think this is necessary? These are basic request and response types of express.

When performing the backend refactoring, we can think about, what part of the code should be documented. What do you think?

**Review by LennartRukower — APPROVED** [review]
Nice work!

**Pull request merged**